Core/TransactionHandler.php

```diff
@@ -0,0 +1,314 @@
+            'amount' => $fAmount,
+            'currency' => $oResponse->getRequestedAmount()->getCurrency(),
+            'responsexml' => base64_encode($oResponse->getRawData()),
+            'date' => $sConvertedTimestamp
+        );
+
+        // add a transaction database entry from properties array
+        Transaction::createDbEntryFromArray($aTransactionProps);
+
+        // check if the parent transaction needs to be set to 'closed'
+        $this->updateParentTransactionStateIfNecessary($oResponse);
+    }
+
+    /**
+     * Checks if the parent transaction still has money to be captured.
+     * If no money is available anymore, the state is set to closed.
+     *
+     * @param SuccessResponse $oResponse
+     *
+     * @since 1.0.1
+     */
+    public function updateParentTransactionStateIfNecessary($oResponse)
+    {
+        $sParentTransactionId = $oResponse->getParentTransactionId();
+
+        $oParentTansaction = oxNew(Transaction::class);
+        $oParentTansaction->loadWithTransactionId($sParentTransactionId);
+
+        $fRestAmount = $this->getTransactionMaxAmount($sParentTransactionId);
+
+        if ($fRestAmount <= 0) {
+            // set transaction state to closed
+            $oParentTansaction->wdoxidee_ordertransactions__state = new Field(Transaction::STATE_CLOSED);
+            $oParentTansaction->save();
+        }
+    }
+
+    /**
+     * Calculates the maximum amount available for a transaction.
+     *
+     * @param string $sTransactionId
+     *
+     * @return float
+     *
+     * @since 1.0.1
+     */
+    public function getTransactionMaxAmount($sTransactionId)
+    {
+        $oTransaction = oxNew(Transaction::class);
+        $oTransaction->loadWithTransactionId($sTransactionId);
+        $fBaseAmount = $oTransaction->wdoxidee_ordertransactions__amount->value;
+
+        // get the child transactions amount summed
+        $oDb = DatabaseProvider::getDb();
+        $oDb->setFetchMode(DatabaseProvider::FETCH_MODE_ASSOC);
+
+        $sDbIdentifier = $oDb->quoteIdentifier('wdoxidee_ordertransactions');
+
+        $sDbQuery = "SELECT SUM(amount) AS childTransactionsTotalAmount FROM {$sDbIdentifier}
+                        WHERE PARENTTRANSACTIONID LIKE ?";
+
+        $aQueryArgs = array($sTransactionId);
+
+        $aResult = $oDb->select($sDbQuery, $aQueryArgs);
+
+        if ($aResult !== false && $aResult->count() > 0) {
+            $fChildAmount = $aResult->fields['childTransactionsTotalAmount'];
+        }
+
+        return $fBaseAmount - $fChildAmount;
+    }
+
+    /**
+     * Recursively calls itself until a transaction without a parent transaction ID is found.
+     * This is the root transaction and its transaction ID is returned.
+     *
+     * @param string $sTransactionId
+     *
+     * @return string root transaction ID
+     *
+     * @since 1.0.1
+     */
+    private function _getRootTransactionId($sTransactionId)
+    {
+        // get transaction to this transaction ID
```

Comment:
We can omit this comment as it is obvious enough.

Reply:
👍